cache/in-memory/src/iter.rs

```diff
@@ -0,0 +1,286 @@
+///
+/// [`std::ops::Deref`] is implemented on this type and derefs to an immutable
+/// reference of the underlying value.
+// We need a separate type from [`Reference`] due to DashMap's iterators
+// returning a different reference type from that of retrieval methods.
+//
+// [`Reference`]: super::Reference
+pub struct IterReference<'a, K, V> {
+    inner: RefMulti<'a, K, V>,
+}
+
+impl<'a, K, V> IterReference<'a, K, V> {
+    /// Create a new iterator element reference.
+    const fn new(inner: RefMulti<'a, K, V>) -> Self {
+        Self { inner }
+    }
+}
+
+impl<K: Eq + Hash, V> IterReference<'_, K, V> {
+    /// Immutable reference to the resource's key.
+    pub fn key(&self) -> &K {
+        self.inner.key()
+    }
+
+    /// Immutable reference to the resource's value.
+    pub fn value(&self) -> &V {
+        self.inner.value()
+    }
+}
+
+impl<K: Eq + Hash, V> Deref for IterReference<'_, K, V> {
+    type Target = V;
+
+    fn deref(&self) -> &Self::Target {
+        self.value()
+    }
+}
+
+/// Interface to create iterators over various resources.
+///
+/// The created iterators will iterate over *all* entities of a resource across
+/// all channels and guilds.
+///
+/// The iteration order of all iterators are arbitrary.
+///
+/// # Examples
+///
+/// Count the number of users in the cache whose username begins with "twi":
+///
+/// ```no_run
+/// use twilight_cache_inmemory::InMemoryCache;
+///
+/// let cache = InMemoryCache::new();
+///
+/// // later in the application...
+/// let count = cache.iter()
+///     .users()
+///     .filter(|user| user.name.starts_with("twi"))
+///     .count();
+///
+/// println!("'twi' users: {}", count);
+/// ```
+///
+/// # Potential inefficiency
+///
+/// Resource iterators over the entire cache are inefficient when the goal is to
+/// iterate over a resource in a specific guild. For example, when performing a
+/// task such as iterating over the members of a specific guild, retrieving the
+/// list of members via [`InMemoryCache::guild_members`] and then calling
+/// [`InMemoryCache::member`] for each item is more efficient. That might look
+/// like:
+///
+/// ```no_run
+/// use twilight_cache_inmemory::InMemoryCache;
+/// use twilight_model::id::GuildId;
+///
+/// let cache = InMemoryCache::new();
+///
+/// // later in the application...
+/// let guild_id = GuildId::new(1).expect("non zero id");
+/// let maybe_guild_members = cache.guild_members(guild_id);
+///
+/// if let Some(guild_members) = maybe_guild_members {
+///     for user_id in guild_members.iter() {
+///         if let Some(member) = cache.member(guild_id, *user_id) {
+///             println!(
+///                 "member id {}'s nickname: {:?}",
+///                 member.user_id(),
+///                 member.nick(),
+///             );
+///         }
+///     }
+/// }
+/// ```
+#[derive(Debug)]
+pub struct InMemoryCacheIter<'a>(&'a InMemoryCache);
+
+impl<'a> InMemoryCacheIter<'a> {
```

Comment:
Could this also have an `into_cache` method, like InMemoryCachePermissions?

Reply:
[`InMemoryCachePermissions::into_cache`](https://docs.rs/twilight-cache-inmemory/0.6.1/twilight_cache_inmemory/permission/struct.InMemoryCachePermissions.html#method.into_cache) was a mistake; it's effectively the same as [`InMemoryCachePermissions::cache_ref`](https://docs.rs/twilight-cache-inmemory/0.6.1/twilight_cache_inmemory/permission/struct.InMemoryCachePermissions.html#method.cache_ref) and needs to be removed.